### ISSUE TITLE
Bug: Composer solver bug in handling of `prefer-stable` sometimes allows update to unstable versions when it shouldn't

### ISSUE THREAD

**Issue by greg-1-anderson** [composer/composer#8882]
My `composer.json`:

```json
{
    "name": "drupal/recommended-project",
    "description": "Project template for Drupal 8 projects with a relocated document root",
    "type": "project",
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        }
    ],
    "require": {
        "composer/installers": "^1.2",
        "drupal/core-composer-scaffold": "^8.8",
        "drupal/core-project-message": "^8.8",
        "drupal/core-recommended": "^8.8"
    },
    "minimum-stability": "dev",
    "prefer-stable": true
}
```
[Full version](https://github.com/drupal/recommended-project/blob/8.8.5/composer.json)

Output of `composer diagnose`:

```
Checking composer.json: OK
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: You are not running the latest snapshot version, run `composer self-update` to update (1.9.0 => ec3f18ee923051e305d3cb87f9d8b6341f17fefc)
Composer version: 1.9.0
PHP version: 7.3.10
PHP binary path: /Users/ganderson/.phpenv/shims/php
```

When I run this command:

```
composer update --dry-run -vvv
```

I get the following output:

```
  - Updating drupal/core-recommended (8.8.4) to drupal/core-recommended (8.9.x-dev 970e22e)
    REASON: Required by the root package: Install command rule (install drupal/core-recommended 8.8.4|install drupal/core-recommended 8.8.0|install drupal/core-recommended 8.8.0-beta1|install drupal/core-recommended 8.8.0-rc1|install drupal/core-recommended 8.8.1|install drupal/core-recommended 8.8.2|install drupal/core-recommended 8.8.3|install drupal/core-recommended 8.8.4|install drupal/core-recommended 8.8.5|install drupal/core-recommended 8.8.x-dev|install drupal/core-recommended 8.9.0-beta1|install drupal/core-recommended 8.9.0-beta2|install drupal/core-recommended 8.9.x-dev)
```

And I expected this to happen:

Composer should update Drupal to 8.8.5, not 8.9.x-dev.

8.9.x-dev is a CANDIDATE because we have `"minimum-stability": "dev"`, but Composer should still prefer 8.8.5, because we have `"prefer-stable": true`. This has been working up to now, but all of a sudden Composer is preferring the dev release over the stable, and no one knows why.

Here's the kicker. If we change to `"minimum-stability": "alpha"` (and keep `"prefer-stable": true`), then we get:

```
composer update --dry-run -vvv

  - Updating drupal/core-recommended (8.8.4) to drupal/core-recommended (8.8.5)
    REASON: Required by the root package: Install command rule (install drupal/core-recommended 8.8.4|install drupal/core-recommended 8.8.0|install drupal/core-recommended 8.8.0-beta1|install drupal/core-recommended 8.8.0-rc1|install drupal/core-recommended 8.8.1|install drupal/core-recommended 8.8.2|install drupal/core-recommended 8.8.3|install drupal/core-recommended 8.8.4|install drupal/core-recommended 8.8.5|install drupal/core-recommended 8.9.0-beta1|install drupal/core-recommended 8.9.0-beta2)
```

So, like the preceding case, it is showing `8.9.0-beta2` as a candidate, but it is picking 8.8.5, like it should due to "prefer-stable".

I am running these tests on Composer 1.9.0, because I know (think?) this worked in the past, and I wanted to remove as many variables as possible to try and figure out what changed, but I can't figure it out.

When I upgrade to Composer 1.10.6 (today's latest stable version), I get the same results.

**Comment by geerlingguy** [comment]
I had the exact same issue (see https://github.com/geerlingguy/jeffgeerling-com/issues/82#issuecomment-625328614).

I was able to work around it by running:

    composer update "drupal/*"

Which updated Drupal to `8.8.5` instead of `8.9.x-dev`. So maybe a conflict from the `drupal/core-recommended` template (something else in it) outside the `drupal/` namespace?

**Comment by greg-1-anderson** [comment]
I just tried the same test after replacing `drupal/core-recommended` with `drupal-core`. The results were correct:

```
composer update --dry-run -vvv

  - Updating drupal/core (8.8.4) to drupal/core (8.8.5)
    REASON: Required by the root package: Install command rule (install drupal/core 8.8.4|install drupal/core 8.8.0|install drupal/core 8.8.0-alpha1|install drupal/core 8.8.0-beta1|install drupal/core 8.8.0-rc1|install drupal/core 8.8.1|install drupal/core 8.8.2|install drupal/core 8.8.3|install drupal/core 8.8.4|install drupal/core 8.8.5|install drupal/core 8.8.x-dev|install drupal/core 8.9.0-beta1|install drupal/core 8.9.0-beta2|install drupal/core 8.9.x-dev)
```

`drupal/core-recommended` is a metapackage that includes `drupal/core`, and also strongly constrains all of the dependencies of `drupal/core` to be the exact versions tested when that version of `drupal/core` was tested and released. It looks like this:

https://github.com/drupal/core-recommended/blob/8.8.4/composer.json

**Comment by greg-1-anderson** [comment]
I also tried dumping data from Packagist:

- curl https://repo.packagist.org/p/drupal/core-recommended.json
- curl https://repo.packagist.org/p/drupal/core.json

Looking at the entries for 8.9.x-dev for both of these finds that they are very similar, save that the first is `"type": "metapackage"`, and the other is not. 

I also tried removing the `vendor` directory before `composer update --dry-run`, but that also did not change the behavior. Even removing the `composer.lock` file and starting fresh gave 8.9.x-dev instead of 8.8.5 stable.

**Comment by greg-1-anderson** [comment]
I was able to reproduce @geerlingguy 's workaround. Here's the odd part, though:

- `composer update 'drupal/*' --dry-run --with-dependencies` ALSO give correct results (updates to 8.8.5)
- `composer update 'drupal/*' --dry-run --with-all-dependencies` gives INCORRECT results (updates to 8.9.x-dev)

**Comment by greg-1-anderson** [comment]
So there is something about composer/installers that is causing this.

That’s the only package that is evaluated with `--with-all-dependencies` that is NOT evaluated with `--with-dependencies` when the whitelist is drupal/*

**Comment by greg-1-anderson** [comment]
So, the other thing that happened recently is we made `drupal/core` require `composer/installers` ^1.9 instead of `composer/installers` ^1.7.  `drupal/core-recommended` is generated from the dependencies of `drupal/core`, so the result of this is that `drupal/core-recommended` requires composer/installers ^1.7 in its 8.8.5 tag, and ^1.9 in its 8.9.x-dev tag.

So it seems that composer/installers is a key component to the behavior that we're seeing here.

Here’s a really big clue.

- `composer update 'drupal/core-recommended' composer/installers --dry-run --with-dependencies` goes to 8.9.x-dev.
- `composer update 'drupal/core-recommended' composer/installers --dry-run` says “Nothing to install or update.”

From this, we can conclude with confidence that something about updating drupal/core-recommended + composer/installers makes drupal/core-recommended 8.8.5 ineligible. If this happens, then Composer will definitely pick 8.9.x-dev, because minimum-stability dev allows it, and prefer-stable true does not stop it when there are not eligible stable releases.

I am unsure how to figure out why this is, though. Adding `-vvv` to the "Nothing to install or update" case above does not print out any additional information about what the solver is doing. Adding `-vvv` to the `--with-dependencies` case doesn't give any information that proves beyond any doubt what the cause is:
```
  - Updating composer/installers (v1.7.0) to composer/installers (v1.9.0)
    REASON: Required by the root package: Install command rule (install composer/installers v1.7.0|install composer/installers v1.7.0|install composer/installers v1.2.0|install composer/installers v1.3.0|install composer/installers v1.4.0|install composer/installers v1.5.0|install composer/installers v1.6.0|install composer/installers v1.7.0|install composer/installers v1.8.0|install composer/installers v1.9.0)

  - Updating drupal/core-recommended (8.8.4) to drupal/core-recommended (8.9.x-dev 970e22e)
    REASON: Required by the root package: Install command rule (install drupal/core-recommended 8.8.4|install drupal/core-recommended 8.8.4|install drupal/core-recommended 8.8.0|install drupal/core-recommended 8.8.0-beta1|install drupal/core-recommended 8.8.0-rc1|install drupal/core-recommended 8.8.1|install drupal/core-recommended 8.8.2|install drupal/core-recommended 8.8.3|install drupal/core-recommended 8.8.4|install drupal/core-recommended 8.8.5|install drupal/core-recommended 8.8.x-dev|install drupal/core-recommended 8.9.0-beta1|install drupal/core-recommended 8.9.0-beta2|install drupal/core-recommended 8.9.x-dev)
```

I think, though, that this is caused by a bug in the Composer solver. Here is my hypothesis:

So Composer sees two valid choices:

- Drupal 8.8.5 (which requires composer installers 1.7.0 exactly, and no other version)
- Drupal 8.9.x-dev (which requires composer installers 1.9.0 exactly, and no other version)

In the workarounds, the second option is NOT viable. Setting minimum stability to “alpha” eliminates 8.9.x from consideration. Keeping composer/installers off the whitelist also prevents 8.9.x-dev from being a candidate, because it must upgrade composer installers at the same time, or not be upgraded.

Of these two choices, both are valid if min stability is “dev”; however, the first should be preferred due to `prefer-stable: true`. If our project is using drupal/core, then the dependency tree is simpler and Composer gets the right answer. With drupal/core-recommended, though, there are two projects being updated, composer installers 1.9.0 is stable, and it NEEDS 8.9.x-dev in order to be upgraded.

I think there’s a bug where the SAT forgives `drupal/core-recommended` for not being stable so that it can satisfy the composer/installers upgrade.

**Comment by greg-1-anderson** [comment]
If the above explanation is in fact correct, then I see only two possible routes forward:

- Fix the Composer 1 solver and make a new release before Drupal 8.9.0 comes out (around 3 June)
- Remove Composer 2 support from Drupal 8.9.0 (this would work because then Drupal 8.9.x would also require composer/installers 1.7.0)

**Comment by greg-1-anderson** [comment]
We might have a better way to implement the second option. If we keep Composer 2 support in Drupal 8.9.0, but pin to composer/installers 1.7.0, then this behavior (early upgrade to 8.9.x-dev) wouldn't happen any longer, but it would still be possible to use Composer 2 with Drupal 8.9.x. In order to do that, you would need to:

- Remove drupal/core-recommended and require drupal/core instead. (This is a normal thing for folks to do to their Drupal sites, if they want to accept the risk and upgrade a dependency sooner than Drupal does.)
- Upgrade your composer/installers to ^1.9

This seems like it would be a reasonable compromise that wouldn't involve changing the Composer 1 solver late in its life cycle.

**Comment by greg-1-anderson** [comment]
I just realized that we can, in fact, test this situation on Composer 2. It is not possible to use Composer 2 with Drupal 8.8.x, because the various plugins it requires do not support the composer api ^2. However, the plugins are only necessary for correct operation; they are not necessary for running `composer update`.

When I run `composer update` on a Drupal 8.8.4 site using Composer 2, this bug still occurs, and Drupal is upgraded to 8.9.x-dev, when it should instead have been updated to 8.8.5.

If I run `composer update 'drupal/core-recommended' --dry-run --with-dependencies -vvv` with Composer 2, then Composer throws an exception:
```
                                
  [ErrorException]              
  Undefined index: drupal/core  
                                

Exception trace:
 () at phar:///Users/ganderson/bin/composer/src/Composer/DependencyResolver/PoolBuilder.php:417
 Composer\Util\ErrorHandler::handle() at phar:///Users/ganderson/bin/composer/src/Composer/DependencyResolver/PoolBuilder.php:417
 Composer\DependencyResolver\PoolBuilder->unfixPackage() at phar:///Users/ganderson/bin/composer/src/Composer/DependencyResolver/PoolBuilder.php:418
 Composer\DependencyResolver\PoolBuilder->unfixPackage() at phar:///Users/ganderson/bin/composer/src/Composer/DependencyResolver/PoolBuilder.php:327
 Composer\DependencyResolver\PoolBuilder->loadPackage() at phar:///Users/ganderson/bin/composer/src/Composer/DependencyResolver/PoolBuilder.php:185
 Composer\DependencyResolver\PoolBuilder->buildPool() at phar:///Users/ganderson/bin/composer/src/Composer/Repository/RepositorySet.php:232
 Composer\Repository\RepositorySet->createPool() at phar:///Users/ganderson/bin/composer/src/Composer/Installer.php:405
 Composer\Installer->doUpdate() at phar:///Users/ganderson/bin/composer/src/Composer/Installer.php:249
 Composer\Installer->run() at phar:///Users/ganderson/bin/composer/src/Composer/Command/UpdateCommand.php:224
 Composer\Command\UpdateCommand->execute() at phar:///Users/ganderson/bin/composer/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at phar:///Users/ganderson/bin/composer/vendor/symfony/console/Application.php:835
 Symfony\Component\Console\Application->doRunCommand() at phar:///Users/ganderson/bin/composer/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at phar:///Users/ganderson/bin/composer/src/Composer/Console/Application.php:287
 Composer\Console\Application->doRun() at phar:///Users/ganderson/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///Users/ganderson/bin/composer/src/Composer/Console/Application.php:113
 Composer\Console\Application->run() at phar:///Users/ganderson/bin/composer/bin/composer:61
 require() at /Users/ganderson/bin/composer:25
```

**Comment by greg-1-anderson** [comment]
Note that the Drupal project has reverted the change to drupal/core-recommended that exposed this bug, so it is no longer possible to reproduce in production.

We have decided that the long-term fix is to remove composer/installers from drupal/core-recommended. This should shield us from this bug for the time being.

Note that the bug in the solver remains, and could potentially affect other project. Also, the Composer 2 bug described above still remains. I will open a separate issue to track that, as it is a separate issue from the solver bug.

**Comment by Seldaek** [comment]
Definitely would be happy to get a stable repro case for the Composer 2 bug as that looks really bad.

Regarding the rest, I imagine the issue is only because you upgraded the composer/installers require 1.7 to 1.9, and it probably decides to pick 1.9 first, then once it has decided that it is looking at the drupal package requiring it and has to pick the dev one as it is the only one which can work with 1.9.

So IMO according to this, it should be safe to do your release as is, because once you have a stable release requiring 1.9, it will go back to safely prefering stable releases when it can, like it always did..

Now obviously I can't really prove this right now, but technically this is not a bug, prefer-stable is just that, a preference. It still came up with a valid solution and that's what matters most.

**Comment by Seldaek** [comment]
Regarding the v2 bug, @naderman I'm guessing it just needs an isset() check at https://github.com/composer/composer/blob/beded906ba1dad53f3a50faea31daf42a6b121b6/src/Composer/DependencyResolver/PoolBuilder.php#L417 - does that sound likely to you?

**Comment by greg-1-anderson** [comment]
Thanks for the feedback, @Seldaek.

I couldn't reproduce the Composer 2 bug; perhaps you already fixed per the suggestion above? I did not check the commit history.

> So IMO according to this, it should be safe to do your release as is, because once you have a stable release requiring 1.9, it will go back to safely prefering stable releases when it can, like it always did..

Unfortunately, things aren't quite so simple. Drupal is very conservative about updating its dependencies, and there is no intention on upgrading the Drupal 8.8 line with composer/installers ^1.9.  That release will only be on the Drupal 8.9 line of releases, and in the mean time, folks who want stable releases want to stay on 8.8 until a stable 8.9 is available. It won't be too long before 8.9.0 is out, but then we're likely to have the same problem again with 9.0 / 9.1.x-dev.

To avoid these problems, we've made an adjustment (removed composer/installers from drupal/core-recommended) to avoid this bug / solver preference in Composer. This is fragile, though; I think that the real solution is to switch to using `minimum stability: stable` in Drupal projects. I created [drupal.org/#3135247](https://www.drupal.org/project/drupal/issues/3135247) to track.

**Comment by naderman** [comment]
> Regarding the v2 bug, @naderman I'm guessing it just needs an isset() check at
> 
> https://github.com/composer/composer/blob/beded906ba1dad53f3a50faea31daf42a6b121b6/src/Composer/DependencyResolver/PoolBuilder.php#L417
> 
> - does that sound likely to you?

Yes entirely correct, since we unset it afterwards, if the same package comes up recursively there it would error.

**Comment by naderman** [comment]
Actually looking at that in some more detail, I'm not sure just adding an isset would appropriately fix this. Since the recursion may then not correctly unfix replacers either? We definitely need a test case for this before we can fix it.

**Comment by greg-1-anderson** [comment]
The bug only happens with very particular options. Here are steps to reproduce:
```
composer create-project drupal/recommended-project:8.8.4
cd recommended-project/
composer update 'drupal/core-recommended' --dry-run --with-dependencies -vvv
```
The project `drupal/core-recommended` is just a metaproject, although it has a lot of dependencies. Since it is a metaproject, it's not possible to run `composer create-project` on it directly (it cannot be a top-level project). I didn't try making a reduced version of `drupal/recommended-project`, although I suspect that a simple project that required only `drupal/core-recommended` would probably reproduce this bug.

**Comment by alexpott** [comment]
@Seldaek thanks for your response. I agree that composer has come up with a valid solution but if a set of completely stable packages is also a valid solution then shouldn't the behaviour of `"prefer-stable": true` be to select that solution?

**Comment by stof** [comment]
`prefer-stable` impacts the selection of versions of a single package. It won't make the solver try to find 2 different object graphs with one more stable than the other (which might involve identifying all the possible solutions to the dependency problem rather than just finding 1 solution to it, which is a lot more costly).

**Comment by stof** [comment]
The way the solver works is that it takes decisions one by one (selecting a version for a package) and moves on until either a solution is found (then things are done) or the resolution blocks. When blocking, it backtracks to the last decision it made, registers an opposite decision (exclude that version of that package) and then tries to take a decision again to find a solution.
This `prefer-stable` config only impacts which version is tried first when taking a decision for a given package. As Jordi said, the decision for `composer/installers` was probably taken first.

**Comment by ryanaslett** [comment]
The only other thing that Im curious about here is if composer shows any preferences for what decisions it goes for with standard packages vs metapackages? Is there any other hinting for decision making priorities? like, would putting composer/installers at the end of the require section vs the beginning have any bearing on what decision tree the sat solver ends up with?

**Comment by stof** [comment]
it might change the order in which rules are processed in the solver, but there is absolutely no guarantee on that (that's internal implementation details, and it might depend on the whole set of rules adding in the project)

**Comment by Seldaek** [comment]
I think this can be fixed at this point. At least I don't see what actionable point there is.. If I missed something please do tell.

**Comment by greg-1-anderson** [comment]
As far as Drupal is concerned, we have moved away from `"minimum-stability": "dev"` + `"prefer-stable": true` as the default for stable projects, which I think is preferable in any event.

I am +1 on any improvements to the handling of prefer-stable that makes the preference stronger. I have no objection to the closing of this issue.
